lsat/store.go

```diff
@@ -50,42 +67,131 @@ func NewFileStore(storeDir string) (*FileStore, error) {
+func (f *FileStore) currentToken() (*Token, error) {
+	switch {
+	case fileExists(f.fileName):
+		return readTokenFile(f.fileName)
+
+	case fileExists(f.fileNamePending):
+		return readTokenFile(f.fileNamePending)
+
+	default:
 		return nil, ErrNoToken
 	}
-	bytes, err := ioutil.ReadFile(f.fileName)
+}
+
+// AllTokens returns all tokens that the store has knowledge of, even if they
+// might be expired. The tokens are mapped by their identifying attribute like
+// file name or storage key.
+//
+// NOTE: This is part of the Store interface.
+func (f *FileStore) AllTokens() (map[string]*Token, error) {
+	tokens := make(map[string]*Token)
+
+	// All tokens start with the same name so we can get them by the prefix.
+	// As the tokens don't expire yet, there currently can't be more than
+	// just one token, either pending or paid.
+	// TODO(guggero): Update comment once tokens expire and we keep backups.
+	tokenDir := filepath.Dir(f.fileName)
+	files, err := ioutil.ReadDir(tokenDir)
 	if err != nil {
 		return nil, err
 	}
-	return deserializeToken(bytes)
+	for _, file := range files {
+		name := file.Name()
+		if !strings.HasPrefix(name, storeFileName) {
+			continue
+		}
+		fileName := filepath.Join(tokenDir, name)
+		token, err := readTokenFile(fileName)
+		if err != nil {
+			return nil, err
+		}
+		tokens[fileName] = token
+	}
+
+	return tokens, nil
 }
 
 // StoreToken saves a token to the store, overwriting any old token if there is
 // one.
 //
 // NOTE: This is part of the Store interface.
-func (f *FileStore) StoreToken(token *Token) error {
-	bytes, err := serializeToken(token)
+func (f *FileStore) StoreToken(newToken *Token) error {
+	// Serialize the token first, before we rename anything.
+	bytes, err := serializeToken(newToken)
 	if err != nil {
 		return err
 	}
-	return ioutil.WriteFile(f.fileName, bytes, 0600)
+
+	// We'll need to know if there is any other token already in place,
+	// either pending or not, that we need to delete or overwrite.
+	currentToken, err := f.currentToken()
+
+	switch {
+	// No token in the store yet, just write it to the corresponding file.
+	case err == ErrNoToken:
+		// What's the target file name we are going to write?
+		newFileName := f.fileName
+		if newToken.isPending() {
+			newFileName = f.fileNamePending
+		}
+		return ioutil.WriteFile(newFileName, bytes, 0600)
+
+	// Fail on any other error.
+	case err != nil:
+		return err
+
+	// Replace a pending token with a paid one.
+	case currentToken.isPending() && !newToken.isPending():
+		// Make sure we replace the the same token, just with a
+		// different state.
+		if currentToken.PaymentHash != newToken.PaymentHash {
+			return fmt.Errorf("new paid token doesn't match " +
+				"existing pending token")
+		}
+
+		// Write the new token first, so we still have the pending
+		// around if something goes wrong.
+		err := ioutil.WriteFile(f.fileName, bytes, 0600)
+		if err != nil {
+			return err
+		}
+		return os.Remove(f.fileNamePending)
```

Comment:
What do you think about not returning the error here? So it's just a best effort operation since we were able to write the token anyway. We can also try removing it on startup within `NewTokenStore` if the file exists and there's a corresponding paid token.